File: LeetCode/C#/0001-TwoSum.cs
/*
Title: 1. Two Sum
Solution:
Difficulty: Easy
Approach: Hash Table for O(n) lookup
Tags: Array, Hash Table
1) Create a dictionary to store numbers and their indices.
2) Iterate through the array with index tracking.
3) For each number, check if (target - number) exists in the dictionary.
4) If complement found, return current index and complement's index.
5) If not found, add current number and its index to the dictionary.
6) Continue until pair is found.

Time Complexity: O(n) where n = nums.length
Space Complexity: O(n) for the hash table
*/
public class Solution {
    public int[] TwoSum(int[] nums, int target)
        {
            int[] result = new int[2];                              // Array to store result indices
            Dictionary<int,int> keyvalue = new Dictionary<int,int>();   // Dictionary to store number -> index mapping
            int index = 0;                                          // Track current index manually
            foreach (int num in nums)                               // Iterate through each number
            {
                if (keyvalue.ContainsKey(target - num))             // If complement exists in dictionary
                {
                    result[0] = index;                              // Set first index to current position
                    result[1] = keyvalue[target - num];             // Set second index to complement's position
                }
                else                                                // If complement not found
                {
                    keyvalue.TryAdd(num,index);                     // Add current number with its index to dictionary
                }

                index++;                                            // Increment index for next iteration
            }
            return result;                                          // Return result with the two indices
        }
}

File: LeetCode/Python/0001-TwoSum.py
'''
Solution: https://leetcode.com/problems/two-sum/solutions/5774602/simplest-solution-python-by-varunve-wapi/
Difficulty: Easy
Approach: Hash Table for O(n) lookup
Tags: Array, Hash Table
1) Create a hashmap to store numbers and their indices.
2) Iterate through the array with index tracking.
3) For each number, calculate the complement (target - number).
4) If complement exists in hashmap, return current index and complement's index.
5) If not found, add current number and its index to the hashmap.
6) Continue until pair is found.

Time Complexity: O(n) where n = len(nums)
Space Complexity: O(n) for the hash table
'''
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashmap = {}                                    # Dictionary to store number -> index mapping
        for i, num in enumerate(nums):                  # Iterate through array with index
            diff = target - num                         # Calculate required complement
            if diff in hashmap:                         # If complement exists in hashmap
                return [i, hashmap[diff]]               # Return current index and complement's index
            hashmap[num] = i                            # Add current number with its index to hashmap
        return []                                       # No solution found (should not happen per problem constraints)